### PR TITLE
Added/improved layout modes

### PR DESCRIPTION
**Pull request by Dylan-Osborne**
Added the "Tiles" layout based on File Explorer (Start to #753)
The tiles layout also supports displaying and editing longer file names.
Here's what it looks like:
Annotation 2020-05-22 200833

Also added support for the current grid layout to be resized as well as switching between layout modes by using ctrl+scroll or ctrl + +\- (#756 point 4)

Here's an example of switching between modes using the ctrl shortcuts:
LayoutModes

When resizing the grid layout using the ctrl shortcuts, icons only refresh their quality/size on a manual refresh. This is just to prevent the performance impacts of reloading the file thumbnails everytime that a user resizes the grid. Thumbnail sizes for the grid view is calculated based on the current grid size to provide a decent quality thumbnail.

One thing that could need improvement would be the layout selection flyout on the bottom right, I've added tiles layout and three general grid sizes (large, medium and small). I had to create icons for the grid sizes, but I'm not quite sure if they fit in, so if anyone wants to redo the icons they are more than welcome to!
Annotation 2020-05-22 195115

### CODE REVIEW

**Comment by Dylan-Osborne** [comment]
If anyone's wondering, the TilesBrower is based on the original PhotoAlbum, so it shares a large portion of the backend code. I decided to keep them as seperate controls as the tiles layout is static while the grid layout has quite a few dynamic properties due to the resize functionality.

**Comment by xPoppyx** [comment]
nice

> icons only refresh their quality/size on a manual refresh. This is just to prevent the performance impacts of reloading the file thumbnails everytime that a user resizes the grid. 

hopefully there is a way to "fix" this to avoid manually refreshing

**Comment by tsvietOK** [comment]
@Dylan-Osborne I don't think it is a good idea to duplicate `PhotoAlbum` class. PhotoAlbum.xaml.cs equals to TilesBrowser.xaml.cs. As i understand difference only in xaml files in Data templates, so why don't to use several data templates in PhotoAlbum.xaml?

**Comment by Dylan-Osborne** [comment]
@tsvietOK Thanks for picking that up, I didn't know you could do that. I'll have a look and see what I can come up with data templates.

@xPoppyx That should be easy enough to do. Currently it's only a once off that you'd need to refresh after resizing the grid, you could also browse to a different directory to achieve the same effect.

**Review by yair100 — APPROVED** [review]

File: Files/BaseLayout.cs
Comment (tsvietOK) [inline Files/BaseLayout.cs:386]: What about "Add" [key](https://docs.microsoft.com/en-us/uwp/api/windows.system.virtualkey?view=winrt-18362)? (code 107)

File: Files/Interacts/Interaction.cs
Comment (tsvietOK) [inline Files/Interacts/Interaction.cs:112]: Don't copy code from PhotoAlbum "if", you can just use 
`if (CurrentSourceType == typeof(PhotoAlbum) || if (CurrentSourceType == typeof(TilesBrowser)))`

**Comment by tsvietOK** [comment]
@Dylan-Osborne examples on how to use different data templates you can find [here](https://github.com/microsoft/Xaml-Controls-Gallery/blob/master/XamlControlsGallery/ControlPages/GridViewPage.xaml) and [here](https://github.com/microsoft/Xaml-Controls-Gallery/blob/master/XamlControlsGallery/ControlPages/GridViewPage.xaml.cs)

File: Files/BaseLayout.cs
Comment (Dylan-Osborne) [reply Files/BaseLayout.cs:386]: As far as I can see, the "Add" key is the + key located on a numeric keypad. Not all keyboards have the numeric keypad, so I chose this route so that the + and - on the main keyboard are used instead.

**Comment by Dylan-Osborne** [comment]
@tsvietOK I've managed to get everything working using data templates. I've had to use a DataTemplateSelector to choose which layout to use, but I've noticed that it executes the selector at least once for each listed item. This is bad for performance, there's even a warning on it's [docs](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/data-template-selector#datatemplateselector-performance-considerations)

Is there a way to select a data template once off instead of for each listed item?
Here's how I've been using the data template so far [xaml](https://github.com/Dylan-Osborne/files-uwp/blob/layout-datatemplate/Files/UserControls/LayoutModes/PhotoAlbum.xaml) and [cs](https://github.com/Dylan-Osborne/files-uwp/blob/layout-datatemplate/Files/UserControls/LayoutModes/PhotoAlbum.xaml.cs)

**Comment by tsvietOK** [comment]
@Dylan-Osborne As i understand, `DataTemplateSelector` is used to define different templates(views) to items at the same time. We are don't need this. Instead use `ItemTemplate` property to set one template for all items (look my example above)

File: Files/BaseLayout.cs
Comment (tsvietOK) [reply Files/BaseLayout.cs:386]: Will it work if you change grid size using numpad keys (+/-)?

Comment (Dylan-Osborne) [reply Files/BaseLayout.cs:382]: I've added support for the numpad keys now👍

File: Files/Interacts/Interaction.cs
Comment (Dylan-Osborne) [reply Files/Interacts/Interaction.cs]: These changes have been fixed/removed

**Comment by tsvietOK** [comment]
@Dylan-Osborne When using dark theme, one icon still black
image

**Comment by tsvietOK** [comment]
@Dylan-Osborne Not necessarily, but you can add ability to translate new strings by adding them into strings-us file:
StatusBarControlTilesView.Text
StatusBarControlGridViewSmall.Text
StatusBarControlGridViewMedium.Text
StatusBarControlGridViewLarge.Text

**Comment by tsvietOK** [comment]
@Dylan-Osborne  For large icon you can use font icon `E71A`

**Comment by tsvietOK** [comment]
I think it is better to add margin at the end of textBlock and textBox 
image

**Comment by tsvietOK** [comment]
And vertically align center name, type and file size

**Comment by tsvietOK** [comment]
And last what i have found, folder icon is pixelated in GridView large moder
image

**Comment by tsvietOK** [comment]
For GridView small you can use font icon E80A.
image

**Comment by tsvietOK** [comment]
@Dylan-Osborne There is a bug. Repro:
Open folder in tiles mode
Change tiles mode to grid view (any size)
Try to open folder context menu
Crash

File: Files/Files.csproj
Comment (yaichenbaum) [inline Files/Files.csproj:270]: Are the Segoe UI assets you can use instead of these?

Comment (Dylan-Osborne) [reply Files/Files.csproj:270]: Yes, I've used the FontIcon fonts @tsvietOK suggested, it looks much better in the dark theme.

I'll commit the changes as soon as I've fixed the folder bug.

**Comment by yaichenbaum** [comment]
@Dylan-Osborne I tested out your changes and I really like them, I noticed two UI issues that should probably be resolved before merging the pull request
- The UI in tiles view is a little too large.
- The file name looks a little too high on folders.

image

**Comment by Dylan-Osborne** [comment]
@yair100 Does the tile sizing look a bit better here? I've reduced the width and height of each cell, I don't want to make the height smaller than this as then it won't be able to wrap longer file names to the second line. I've also vertically centered the text details.
image

**Comment by duke7553** [comment]
@Dylan-Osborne Does the new Tiles layout support lazy loading of extended properties like file type, icon, and size?

**Comment by yaichenbaum** [comment]
@Dylan-Osborne That looks amazing 🙂. Is there anything else we are still waiting on for this pull request to be merged @tsvietOK @duke7553?

**Review by yair100 — APPROVED** [review]

**Comment by tsvietOK** [comment]
@yair100 Wait a bit, as i understand [this](https://github.com/duke7553/files-uwp/pull/827#issuecomment-633101955) bug still not fixed

File: Files/View Models/ItemViewModel.cs
Comment (duke7553) [inline Files/View Models/ItemViewModel.cs:612]: We actually want to use `ThumbnailMode.SingleItem` here for more descriptive thumbnails.

**Comment by yaichenbaum** [comment]
@tsvietOK I wasn't able to reproduce that bug on my end.

**Comment by Dylan-Osborne** [comment]
@duke7553 It uses the same backend as the original PhotoAlbum, so it should support lazy loading.
@tsvietOK I've fixed the bug, just haven't committed the changes yet.

**Review by yair100 — APPROVED** [review]

**Comment by yaichenbaum** [comment]
While we are working on this, can you rename PhotoView to GridView?

**Comment by tsvietOK** [comment]
@yair100 What about `GridViewBrowser`? Because `GridView` is a component of Microsoft.UI.Xaml.Controls

File: Files/UserControls/LayoutModes/GridView.xaml
Comment (yaichenbaum) [inline Files/UserControls/LayoutModes/GridView.xaml:2]: This should be GridViewBrowser.

File: Files/UserControls/LayoutModes/GridView.xaml.cs
Comment (yaichenbaum) [inline Files/UserControls/LayoutModes/GridView.xaml.cs:14]: This should be GridViewBrowser.

**Review by yair100 — CHANGES_REQUESTED** [review]

File: Files/UserControls/LayoutModes/GridView.xaml.cs
Comment (Dylan-Osborne) [reply Files/UserControls/LayoutModes/GridView.xaml.cs]: Thanks for picking this up, it should be fixed now

File: Files/View Models/ItemViewModel.cs
Comment (duke7553) [reply Files/View Models/ItemViewModel.cs]: @Dylan-Osborne

**Review by lukeblevins — APPROVED** [review]
Very happy to see this change

**Review by yair100 — APPROVED** [review]

**Comment by yaichenbaum** [comment]
@Dylan-Osborne Thank you so much for this pull request, this has been merged into master and will roll out with the next release.

**Comment by xPoppyx** [comment]
in grid view (small, medium and large) the images with more height than width are cut off, unable to show the name
Anotación 2020-05-24 150145

**Comment by xPoppyx** [comment]
in tiles view the file format is not shown with a correct translation
Anotación 2020-05-25 223346

**Comment by yaichenbaum** [comment]
@xPoppyx Can you please open a separate issue for that?

**Pull request merged**
